src/model/sample.rs

```diff
@@ -608,13 +608,43 @@ mod tests {
     use bio::io::fasta;
 
 
+    fn recode_evidence(string: String) -> Evidence {
+        match &*string.to_string() {
+            "Alignment" => Evidence::Alignment(string),
+            _ => Evidence::InsertSize(string)
+        }
+    }
+
+    #[derive(Debug, Deserialize)]
+    struct Row {
+        chrom: String,
+        pos: u32,
+        allele: u32,
+        sample: String,
+        prob_mapping: LogProb,
+        prob_alt: LogProb,
+        prob_ref: LogProb,
+        prob_mismapped: LogProb,
+        evidence: String
+    }
+
     fn read_observations(path: &str) -> Vec<Observation> {
-        let mut reader = csv::Reader::from_file(path).expect("error reading example").delimiter(b'\t');
-        let obs = reader.decode().collect::<Result<Vec<(String, u32, u32, String, Observation)>, _>>().unwrap();
-        let groups = obs.into_iter().group_by(|&(_, _, _, ref sample, _)| {
-            sample == "case"
-        });
-        let case_obs = groups.into_iter().next().unwrap().1.into_iter().map(|(_, _, _, _, obs)| obs).collect_vec();
+        let mut reader = csv::ReaderBuilder::new().delimiter(b'\t').from_path(path).expect("error reading example");
+        let mut case_obs = vec![];
+        for row in reader.deserialize() {
```

Comment:
I would prefer to still deserialize into `(String, u32, u32, String, Observation)`. Is that possible?

Reply:
When trying to unpack as similar to the previous code as possible, i.e. with:
```
let unpacked = reader.deserialize().collect::<Result<Vec<(String, u32, u32, String, Observation)>, _>>().unwrap();
```

It panics with:
```
'called `Result::unwrap()` on an `Err` value: Error(Deserialize { pos: Some(Position { byte: 79, line: 2, record: 1 }), err: DeserializeError { field: Some(8), kind: ParseFloat(ParseFloatError { kind: Invalid }) } })', /checkout/src/libcore/result.rs:859
```

The documentation indicates, that automatic reading into structs will only work when `has_headers(false)` is set for the `ReaderBuilder`. However, no other method to skip the first line in reading is provided, so we can only do avoid choking on the first line on the iterator level, e.g. like this:

```
let mut reader = csv::ReaderBuilder::new().has_headers(false).delimiter(b'\t').from_path(path).expect("error reading example");
let deserialized = reader.deserialize().skip(1);
let unwrapped = deserialized.into_iter().collect::<Result<Vec<(String, u32, u32, String, Observation)>, _>>().unwrap();
```

But this still chokes on the unwrap, with a different error:
```
panicked at 'called `Result::unwrap()` on an `Err` value: Error(Deserialize { pos: Some(Position { byte: 79, line: 2, record: 1 }), err: DeserializeError { field: None, kind: Message("invalid type: unit variant, expected newtype variant") } })'
```

I'm assuming, this has to do with the `enum Evidence` not being tagged, because this Error goes away when I annotate the `enum` with:
```
#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(untagged)]
pub enum Evidence {
    /// Insert size of fragment
    InsertSize(String),
    /// Alignment of a single read
    Alignment(String)
}
```

However, this then chokes on the first line with a number for an`InsertSize` in it instead of the String `Alignment` with:
```
panicked at 'called `Result::unwrap()` on an `Err` value: Error(Deserialize { pos: Some(Position { byte: 3748, line: 51, record: 50 }), err: DeserializeError { field: None, kind: Message("data did not match any variant of untagged enum Evidence") } })'
```

And I'm not sure where to go with that... No clue how to give it a hint about the type based on what it reads from the file, there. Any suggestions?

Reply:
Ok, I see. This is crap. I suggest we do the following.
1. We keep the workaround with `Row`. It is not so much of an issue, because it is only in the test cases.
2. If you find the time, you could maybe create a minimal example and post an issue in the csv github repo, such that the author becomes aware of it.

Reply:
Yep, I'll try to create that minimal example tomorrow in the morning. Tuple or struct containing a struct that contains such an ambiguous enum and the corresponding csv file to go with it.

Reply:
Just for cross-reference, issues raised upstream:
https://github.com/BurntSushi/rust-csv/issues/100
https://github.com/BurntSushi/rust-csv/issues/101

Feel free to chime in or correct me over there! ;)

Reply:
Thanks! All fine!

Reply:
1. It looks like automatic deserializing of `csv` input previously written out from `csv`-serialized version of our `enum Evidence` will not be supported (see discussion in <https://github.com/BurntSushi/rust-csv/issues/100>). With the current functionality, we could get to a `csv`-supported version by making the variant arguments distinct types, e.g. making it:
```
pub enum Evidence {
    InsertSize(f64),
    Alignment(String)
}
```

2. Once this is sorted, we could get around the header matching that does not work in the Tuple `.collect()` encompassing both primitives and the `struct Observation` by using `.has_headers(false)` in the `csv::ReaderBuilder` and a `.skip(1)` after the `.deserialize()` (see discussion in <https://github.com/BurntSushi/rust-csv/issues/101>).

With 1. we avoid the current use of the helper `fn recode_evidence(string: String) -> Evidence` and with 2. we would get around defining the helper `struct Row`. But this would involve changing `enum Evidence`, and I'm not sure if that would be desirable--although I don't really see, why it currently is `InsertSize(String)`, when the variant argument should usually be a number? Should we make this change, or leave the code as is?